### PR TITLE
v4.2: switch custom form control

### PR DESCRIPTION
**Pull request by gijsbotje**
This adds a neat styling feature for the custom form controls.
Markup stays the same as the checkbox but displays it as a toggle switch.

schermafbeelding 2017-07-06 om 11 25 55

https://codepen.io/gijsbotje/pen/rwKMmy

https://deploy-preview-23004--twbs-bootstrap4.netlify.com/docs/4.1/components/forms/#switches

### CODE REVIEW

File: docs/4.0/components/forms.md
Comment (skmedix) [inline docs/4.0/components/forms.md]: First letter should be capitalized

**Review by skmedix — CHANGES_REQUESTED** [review]

File: docs/4.0/components/forms.md
Comment (gijsbotje) [inline docs/4.0/components/forms.md]: fixed

**Comment by gijsbotje** [comment]
fixes #22090

**Comment by mdo** [comment]
Dope! Thank you for jumping in on this. I'll take a look at it for our v4.1 ship :).

**Comment by gijsbotje** [comment]
@mdo @XhmikosR @andresgalante I messed up here :( and cant figure out how to revert it, any suggestions?
Sorry for this... don't understand what wen't wrong

**Comment by XhmikosR** [comment]
@gijsbotje: you can just do `git rebase -i upstream/v4-dev` and clean up your local branch.

**Comment by gijsbotje** [comment]
maybe i did something wrong, but it didn't work...

**Comment by XhmikosR** [comment]
@gijsbotje: you had messed up your branch. Do not merge from upstream. Just rebase it, if needed.

Now, I cleaned it up and force pushed it. So, make sure you do `git fetch origin && git reset --hard origin/custom-switch-form-control` before anything else otherwise you will overwrite the cleanup.

**Comment by gijsbotje** [comment]
@XhmikosR thanks dude, will do
i tried to make a pr for a different issue and my forked v4-dev branch is messed up i think
i'll see if i can fix it all

File: docs/4.0/components/forms.md
Comment (patrickhlauke) [inline docs/4.0/components/forms.md]: Probably want to uppercase the "Toggle" text

Comment (gijsbotje) [reply docs/4.0/components/forms.md]: fixed :)

**Comment by m5o** [comment]
@gijsbotje / @mdo the solution to modifying the custom-checkbox with `.custom-checkbox-switch` is super handy. 
http://wingman.mediumra.re/components-bootstrap.html#forms

File: docs/4.0/components/forms.md
Comment (m5o) [inline docs/4.0/components/forms.md]: `.custom-switch-animated` is not defined.

Comment (gijsbotje) [reply docs/4.0/components/forms.md]: have a look at https://github.com/twbs/bootstrap/pull/23004/files#diff-e5540857a54676b43146ad92add2dbdeR182
it's actually is defined as a nested sass selector

Comment (wolfy1339) [reply docs/4.0/components/forms.md]: They don't like that kind of stuff. Don't use nesting to create new classes

**Review by ayoubkhan558-zz — APPROVED** [review]
I think it should be added to bootstrap, so that we don't need any external library for this purpose.

**Review by ayoubkhan558-zz — CHANGES_REQUESTED** [review]
Its a good option to have

**Comment by claudioalmeiida** [comment]
"mdo removed this from Inbox in v4.2 13 days ago" :(

**Comment by guledali** [comment]
@mdo  @XhmikosR Any chance we could see the toggle-switch in this release?

File: scss/_custom-forms.scss
Comment (MartijnCuppens) [inline scss/_custom-forms.scss]: I think we don't need a animated and non animated version, you may use the transition in the default version. Also, use the `transition` mixin for this.

**Comment by XhmikosR** [comment]
Personally I'm not sure it's ready yet.

File: scss/_variables.scss
Comment (MartijnCuppens) [inline scss/_variables.scss]: It's possible to do this without the base64 icon, just use `border-radius` for the element

File: scss/_custom-forms.scss
Comment (MartijnCuppens) [inline scss/_custom-forms.scss]: Don't use background-position for this, use the transform property, this is more performant.

**Review by MartijnCuppens — DISMISSED** [review]

**Comment by XhmikosR** [comment]
@gijsbotje: will you be able to address @MartijnCuppens's comments above?

**Comment by gijsbotje** [comment]
sorry got pre-occupied with work. i'll try and fix it before the weekend

**Comment by gijsbotje** [comment]
i'm struggling a little bit...
as the colour scheme changed, the unchecked state of the switch now doesn't have a gray background. which results into the indicator being not visible until the switch is checked.

to illustrate my issue:
image

@MartijnCuppens @XhmikosR any suggestions on what colors to use for what?

**Comment by XhmikosR** [comment]
Yeah, that was the issue I noticed too when I said I didn't think this was ready along with transitioning costly properties.

I'll try to have a look tomorrow, but @MartijnCuppens will probably beat me to it.

**Comment by gijsbotje** [comment]
would this be ok? i used the color of the border for the indicator
image

**Comment by rafalp** [comment]
Can you include screenshot of other special control (eg. checkbox) bein on/off/disabled?

Also, is there "disabled + on" state?

**Comment by gijsbotje** [comment]
@rafalp i've paired the switch with its checkbox equivalent
image

thanks for reminding me about the `:checked:disabled` state, forgot about that

**Comment by gijsbotje** [comment]
@MartijnCuppens fixed your comments as best as I could think of.
main question to resolve is if the color for the circle when it's unchecked is ok

**Comment by rafalp** [comment]
Those are looking very good, thanks 👍

File: scss/_custom-forms.scss
Comment (MartijnCuppens) [inline scss/_custom-forms.scss]: This is always `50%` I think?

File: scss/_custom-forms.scss
Comment (MartijnCuppens) [inline scss/_custom-forms.scss]: `background-size` is only relevant for background images which we don't have.

**Review by MartijnCuppens — DISMISSED** [review]

**Comment by MartijnCuppens** [comment]
@gijsbotje, the color of the circle looks ok.
IMO the switch looks a bit small, maybe we should try to increase the size a bit? 
It would also be nice to have total control over the track width and height and the spacing between the switch and the label text. With 'the track', I mean this element:
image

Maybe it's possible to tweak this a bit to so that it's possible to set the following variables:
- The track width
- The track height
- Margin between the switch and the label text
- The size of the circle

If we have control over this in `_variables.scss`, it would be easier to control the look of the switch and it would also be possible to make them look like in this PR:
https://github.com/twbs/bootstrap/pull/26691

File: scss/_custom-forms.scss
Comment (MartijnCuppens) [reply scss/_custom-forms.scss]: Forget this, you actually also took care of the ability to use switches that aren't rounded:
image

Sweet!

**Comment by rafalp** [comment]
I would be careful with making switch too big - you'll potentially end with component being too big to use "inline", requiring special positioning or, worse, custom form layout altogether.

**Comment by MartijnCuppens** [comment]
Good point, @rafalp. Forget about size increment than.

**Comment by rafalp** [comment]
Yeah, I am thinking now how this component should be sized. Right now its sized same way checkbox is, meaning its interchangeable with it. Perhaps it should also support `.form-control-lg` and `.form-control-sm` for extra variety?

Just dropping ideas - it may be too late for this, or material for new PR bumping extra sizing to custom controls?

**Comment by gijsbotje** [comment]
> 💡 Support for contextual classes could be interesting, too.

I would do that as a seperate PR as it also affects the other controls

 

> Yeah, I am thinking now how this component should be sized. Right now its sized same way checkbox is, meaning its interchangeable with it. Perhaps it should also support `.form-control-lg` and `.form-control-sm` for extra variety?
> 
> Just dropping ideas - it may be too late for this, or material for new PR bumping extra sizing to custom controls?

Indeed as you said yourself `material for new PR` 😄 
 

> Maybe it's possible to tweak this a bit to so that it's possible to set the following variables:
> 
> - The track width
> - The track height
> - Margin between the switch and the label text
> - The size of the circle
> If we have control over this in _variables.scss, it would be easier to control the look of the switch and it would also be possible to make them look like in this PR:
> #26691 

I was fiddling around with the variable `custom-control-gutter`, but this isn't really the gutter, right? It doesn't define the space between the indicator and the label, but rather defines the space that the label is pushed over. Then the indicator is pulled back the same amount.
So by default the gutter is set to 1.5rem. This gives the checkbox and radio a breathing room of .5rem. Therefore shouldn't gutter be .5rem? And the left properties set to `$custom-control-indicator-size + $custom-control-gutter` which would result in 1.5rem?

If we implement this change, controlling `the track width` and `space between switch and the label` would be much easier. to get it to be accurate in the current setup i would need to do something like:
`$custom-switch-width + ($custom-control-gutter - $custom-control-indicator-size)`
So if you guys agree with the change to `$custom-control-gutter` I will make a separate PR with those changes. I can imagine this would have some impact on people that make use of the sass, so that would be your call.

What i'm really struggling with and get wrap my head around is making the height and circle size customizable. I'm currently using the calc css function to get the height and with like this:
`calc(#{$custom-control-indicator-size} - #{$custom-control-indicator-border-width * 4})`
First the 1rem indicator-size then subtracting the border width * 4 to create space between the circle and the track. This can not be done in sass math as `$custom-control-indicator-border-width` is a pixel value and `$custom-control-indicator-size` is a rem value, which are incompatible in sass.

i'm also struggling with the positioning of the circle when you increase or decrease the size of the circle. currently this calc function is taking care of the vertical positioning:
`calc(#{(($font-size-base * $line-height-base - $custom-control-indicator-size) / 2)} + #{$custom-control-indicator-border-width * 2})`

I'm far from a math guy so if so if someone could assist me in these issues that would be much appreciated.

**Review by MartijnCuppens — APPROVED** [review]
LGTM, more control over sizing is something we can look into later

**Pull request merged**
